File: SideCord/Panel/PanelGeometry.swift
import AppKit

enum PanelGeometry {
    static let minimumWidth: CGFloat = 320
    static let maximumDisplayFraction: CGFloat = 0.8
    static let hiddenOvershoot: CGFloat = 4
    static let railWidth: CGFloat = 76
    static let railGap: CGFloat = 12
    static let railVerticalInset: CGFloat = 12
    static let attentionGlowWidth: CGFloat = 72
    static let incomingCallCardSize = NSSize(width: 292, height: 92)
    static let incomingCallCardInset: CGFloat = 14
    static let onboardingCompanionSize = NSSize(width: 430, height: 650)
    static let onboardingCompanionGap: CGFloat = 18

    static func displayID(for screen: NSScreen) -> String {
        let key = NSDeviceDescriptionKey("NSScreenNumber")
        if let number = screen.deviceDescription[key] as? NSNumber {
            return number.stringValue
        }
        return String(describing: screen)
    }

    static func constrainedWidth(
        _ requestedWidth: CGFloat,
        in usableFrame: NSRect,
        inset: CGFloat = 0
    ) -> CGFloat {
        let safeInset = constrainedInset(inset, in: usableFrame)
        let insetAvailableWidth = max(1, usableFrame.width - (safeInset * 2))
        let maximum = max(
            1,
            min(usableFrame.width * maximumDisplayFraction, insetAvailableWidth)
        )
        let minimum = min(minimumWidth, maximum)
        guard requestedWidth.isFinite else {
            return min(420, maximum)
        }
        return min(max(requestedWidth, minimum), maximum)
    }

    static func sidebarFrame(
        in usableFrame: NSRect,
        edge: SidebarEdge,
        requestedWidth: CGFloat,
        inset: CGFloat = 0
    ) -> NSRect {
        let safeInset = constrainedInset(inset, in: usableFrame)
        let width = constrainedWidth(requestedWidth, in: usableFrame, inset: safeInset)
        let x = edge == .left
            ? usableFrame.minX + safeInset
            : usableFrame.maxX - width - safeInset
        return NSRect(
            x: x,
            y: usableFrame.minY + safeInset,
            width: width,
            height: max(1, usableFrame.height - (safeInset * 2))
        )
    }

    static func railFrame(
        adjacentTo sidebarFrame: NSRect,
        in usableFrame: NSRect,
        edge: SidebarEdge,
        requestedWidth: CGFloat = railWidth,
        gap: CGFloat = railGap,
        verticalInset: CGFloat = railVerticalInset
    ) -> NSRect? {
        guard !sidebarFrame.isEmpty,
              requestedWidth.isFinite,
              gap.isFinite,
              verticalInset.isFinite
        else { return nil }

        let safeGap = max(0, gap)
        let availableWidth: CGFloat
        let x: CGFloat
        switch edge {
        case .right:
            availableWidth = sidebarFrame.minX - safeGap - usableFrame.minX
            let width = min(max(0, requestedWidth), max(0, availableWidth))
            guard width >= 1 else { return nil }
            x = sidebarFrame.minX - safeGap - width
        case .left:
            availableWidth = usableFrame.maxX - sidebarFrame.maxX - safeGap
            let width = min(max(0, requestedWidth), max(0, availableWidth))
            guard width >= 1 else { return nil }
            x = sidebarFrame.maxX + safeGap
        }

        let width = min(max(0, requestedWidth), max(0, availableWidth))
        let safeVerticalInset = min(
            max(0, verticalInset),
            max(0, (sidebarFrame.height - 1) / 2)
        )
        let minY = max(usableFrame.minY, sidebarFrame.minY + safeVerticalInset)
        let maxY = min(usableFrame.maxY, sidebarFrame.maxY - safeVerticalInset)
        guard maxY > minY else { return nil }

        return NSRect(x: x, y: minY, width: width, height: maxY - minY)
    }

    static func hiddenFrame(
        from visibleFrame: NSRect,
        screenFrame: NSRect,
        edge: SidebarEdge
    ) -> NSRect {
        var frame = visibleFrame
        switch edge {
        case .left:
            frame.origin.x = screenFrame.minX - visibleFrame.width - hiddenOvershoot
        case .right:
            frame.origin.x = screenFrame.maxX + hiddenOvershoot
        }
        return frame
    }

    static func attentionGlowFrame(
        in screenFrame: NSRect,
        edge: SidebarEdge,
        requestedWidth: CGFloat = attentionGlowWidth
    ) -> NSRect {
        let safeWidth = requestedWidth.isFinite
            ? min(max(1, requestedWidth), max(1, screenFrame.width))
            : min(attentionGlowWidth, max(1, screenFrame.width))
        return NSRect(
            x: edge == .left ? screenFrame.minX : screenFrame.maxX - safeWidth,
            y: screenFrame.minY,
            width: safeWidth,
            height: max(1, screenFrame.height)
        )
    }

    static func incomingCallCardFrame(
        in screenFrame: NSRect,
        edge: SidebarEdge,
        size: NSSize = incomingCallCardSize,
        inset: CGFloat = incomingCallCardInset
    ) -> NSRect {
        let width = min(max(1, size.width), max(1, screenFrame.width))
        let height = min(max(1, size.height), max(1, screenFrame.height))
        let safeInset = min(max(0, inset), max(0, screenFrame.width - width))
        return NSRect(
            x: edge == .left
                ? screenFrame.minX + safeInset
                : screenFrame.maxX - width - safeInset,
            y: screenFrame.midY - (height / 2),
            width: width,
            height: height
        )
    }

    static func onboardingCompanionFrame(
        adjacentTo sidebarFrame: NSRect,
        in usableFrame: NSRect,
        edge: SidebarEdge,
        requestedSize: NSSize = onboardingCompanionSize
    ) -> NSRect {
        let availableWidth = max(1, usableFrame.width - 24)
        let availableHeight = max(1, usableFrame.height - 24)
        let requestedWidth = requestedSize.width.isFinite
            ? requestedSize.width
            : onboardingCompanionSize.width
        let requestedHeight = requestedSize.height.isFinite
            ? requestedSize.height
            : onboardingCompanionSize.height
        let width = min(max(1, requestedWidth), availableWidth)
        let height = min(max(1, requestedHeight), availableHeight)
        let preferredX = edge == .right
            ? sidebarFrame.minX - onboardingCompanionGap - width
            : sidebarFrame.maxX + onboardingCompanionGap
        let fallbackX = edge == .right
            ? sidebarFrame.minX - (width * 0.72)
            : sidebarFrame.maxX - (width * 0.28)
        let x = min(
            max(preferredX, usableFrame.minX + 12),
            usableFrame.maxX - width - 12
        )
        let resolvedX = sidebarFrame.intersects(
            NSRect(x: x, y: sidebarFrame.midY - height / 2, width: width, height: height)
        ) ? min(max(fallbackX, usableFrame.minX + 12), usableFrame.maxX - width - 12) : x
        return NSRect(
            x: resolvedX,
            y: min(
                max(sidebarFrame.midY - height / 2, usableFrame.minY + 12),
                usableFrame.maxY - height - 12
            ),
            width: width,
            height: height
        )
    }

    static func screen(containing point: NSPoint, from screens: [NSScreen]) -> NSScreen? {
        if let containingScreen = screens.first(where: { NSMouseInRect(point, $0.frame, false) }) {
            return containingScreen
        }

        return screens.min { lhs, rhs in
            squaredDistance(from: point, to: lhs.frame)
                < squaredDistance(from: point, to: rhs.frame)
        }
    }

    static func isEdgeExposed(
        of screen: NSScreen,
        edge: SidebarEdge,
        atY y: CGFloat,
        among screens: [NSScreen]
    ) -> Bool {
        let sampleX = edge == .left ? screen.frame.minX - 1 : screen.frame.maxX + 1
        let sample = NSPoint(x: sampleX, y: y)
        return !screens.contains { candidate in
            candidate !== screen && NSMouseInRect(sample, candidate.frame, false)
        }
    }

    private static func squaredDistance(from point: NSPoint, to rect: NSRect) -> CGFloat {
        let dx = max(max(rect.minX - point.x, 0), point.x - rect.maxX)
        let dy = max(max(rect.minY - point.y, 0), point.y - rect.maxY)
        return dx * dx + dy * dy
    }

    private static func constrainedInset(_ requestedInset: CGFloat, in usableFrame: NSRect) -> CGFloat {
        guard requestedInset.isFinite else { return 0 }
        let maximum = max(0, (min(usableFrame.width, usableFrame.height) - 1) / 2)
        return min(max(requestedInset, 0), maximum)
    }

}

File: Windows/src/main/panel-geometry.js
'use strict';

const MINIMUM_WIDTH = 320;
const MAXIMUM_DISPLAY_FRACTION = 0.8;
const HIDDEN_OVERSHOOT = 4;
const RAIL_WIDTH = 76;
const RAIL_GAP = 12;
const RAIL_VERTICAL_INSET = 12;

function finite(value, fallback) {
  return Number.isFinite(value) ? value : fallback;
}

function constrainedInset(requestedInset, workArea) {
  const maximum = Math.max(0, (Math.min(workArea.width, workArea.height) - 1) / 2);
  return Math.min(Math.max(finite(requestedInset, 0), 0), maximum);
}

function constrainedWidth(requestedWidth, workArea, inset = 0) {
  const safeInset = constrainedInset(inset, workArea);
  const insetAvailableWidth = Math.max(1, workArea.width - safeInset * 2);
  const maximum = Math.max(1, Math.min(workArea.width * MAXIMUM_DISPLAY_FRACTION, insetAvailableWidth));
  const minimum = Math.min(MINIMUM_WIDTH, maximum);
  return Math.min(Math.max(finite(requestedWidth, 420), minimum), maximum);
}

function sidebarBounds(workArea, edge, requestedWidth, inset = 0) {
  const safeInset = constrainedInset(inset, workArea);
  const width = Math.round(constrainedWidth(requestedWidth, workArea, safeInset));
  const x = edge === 'left'
    ? workArea.x + safeInset
    : workArea.x + workArea.width - width - safeInset;
  return {
    x: Math.round(x),
    y: Math.round(workArea.y + safeInset),
    width,
    height: Math.max(1, Math.round(workArea.height - safeInset * 2))
  };
}

function hiddenBounds(visibleBounds, displayBounds, edge) {
  return {
    ...visibleBounds,
    x: Math.round(edge === 'left'
      ? displayBounds.x - visibleBounds.width - HIDDEN_OVERSHOOT
      : displayBounds.x + displayBounds.width + HIDDEN_OVERSHOOT)
  };
}

function glowBounds(displayBounds, edge, requestedWidth = 72) {
  const width = Math.round(Math.min(Math.max(1, finite(requestedWidth, 72)), Math.max(1, displayBounds.width)));
  return {
    x: edge === 'left' ? displayBounds.x : displayBounds.x + displayBounds.width - width,
    y: displayBounds.y,
    width,
    height: Math.max(1, displayBounds.height)
  };
}

function railBounds(sidebar, workArea, edge, requestedWidth = RAIL_WIDTH, gap = RAIL_GAP) {
  if (!sidebar || sidebar.width < 1 || sidebar.height < 1) return null;
  const safeGap = Math.max(0, finite(gap, 0));
  const availableWidth = edge === 'right'
    ? sidebar.x - safeGap - workArea.x
    : workArea.x + workArea.width - (sidebar.x + sidebar.width) - safeGap;
  const width = Math.min(Math.max(0, finite(requestedWidth, RAIL_WIDTH)), Math.max(0, availableWidth));
  if (width < 1) return null;
  const inset = Math.min(RAIL_VERTICAL_INSET, Math.max(0, (sidebar.height - 1) / 2));
  const y = Math.max(workArea.y, sidebar.y + inset);
  const bottom = Math.min(workArea.y + workArea.height, sidebar.y + sidebar.height - inset);
  if (bottom <= y) return null;
  return {
    x: Math.round(edge === 'right' ? sidebar.x - safeGap - width : sidebar.x + sidebar.width + safeGap),
    y: Math.round(y), width: Math.round(width), height: Math.round(bottom - y)
  };
}

function onboardingCompanionBounds(sidebar, workArea, edge, size = { width: 430, height: 650 }) {
  const width = Math.min(Math.max(1, finite(size.width, 430)), Math.max(1, workArea.width - 24));
  const height = Math.min(Math.max(1, finite(size.height, 650)), Math.max(1, workArea.height - 24));
  const preferredX = edge === 'right'
    ? sidebar.x - 18 - width
    : sidebar.x + sidebar.width + 18;
  const fallbackX = edge === 'right'
    ? sidebar.x - width * 0.72
    : sidebar.x + sidebar.width - width * 0.28;
  const clampX = (x) => Math.min(Math.max(x, workArea.x + 12), workArea.x + workArea.width - width - 12);
  let x = clampX(preferredX);
  const y = Math.min(
    Math.max(sidebar.y + (sidebar.height - height) / 2, workArea.y + 12),
    workArea.y + workArea.height - height - 12
  );
  const intersects = x < sidebar.x + sidebar.width && x + width > sidebar.x
    && y < sidebar.y + sidebar.height && y + height > sidebar.y;
  if (intersects) x = clampX(fallbackX);
  return { x: Math.round(x), y: Math.round(y), width: Math.round(width), height: Math.round(height) };
}

function containsPoint(bounds, point, padding = 0) {
  return point.x >= bounds.x - padding
    && point.x <= bounds.x + bounds.width + padding
    && point.y >= bounds.y - padding
    && point.y <= bounds.y + bounds.height + padding;
}

function isEdgeExposed(display, edge, y, displays) {
  const sample = {
    x: edge === 'left' ? display.bounds.x - 1 : display.bounds.x + display.bounds.width + 1,
    y
  };
  return !displays.some((candidate) => candidate.id !== display.id && containsPoint(candidate.bounds, sample));
}

module.exports = {
  MINIMUM_WIDTH,
  MAXIMUM_DISPLAY_FRACTION,
  RAIL_WIDTH,
  RAIL_GAP,
  constrainedInset,
  constrainedWidth,
  sidebarBounds,
  hiddenBounds,
  glowBounds,
  railBounds,
  onboardingCompanionBounds,
  containsPoint,
  isEdgeExposed
};
